pkg/alertmanager/distributor.go

```diff
@@ -254,14 +254,32 @@ func (d *Distributor) doUnary(userID string, w http.ResponseWriter, r *http.Requ
 	defer sp.Finish()
 	// Until we have a mechanism to combine the results from multiple alertmanagers,
 	// we forward the request to only only of the alertmanagers.
-	amDesc := replicationSet.Instances[rand.Intn(len(replicationSet.Instances))]
-	resp, err := d.doRequest(ctx, amDesc, req)
-	if err != nil {
-		respondFromError(err, w, logger)
-		return
-	}
+	instances := replicationSet.Instances
+
+	// Randomize the list of instances to not always query the same one.
+	rand.Shuffle(len(instances), func(i, j int) {
+		instances[i], instances[j] = instances[j], instances[i]
+	})
+
+	for i, instance := range instances {
+		resp, err := d.doRequest(ctx, instance, req)
+		// Only return error if there is no more instances to try
+		if err != nil && i == len(instances)-1 {
```

Comment:
hummm..

i don't think this is doing what u wanna do.

We should retry only on "GET" requests? but this condition will be `true` only on the last instance and... we will keep trying (or retrying) on all instances.. no? Am i missing something?

Reply:
yes,  I missed a check there. Modified the logic to address this.